x/feedistribution/keeper/allocation.go

```diff
@@ -0,0 +1,171 @@
+package keeper
+
+import (
+	"sort"
+
+	"cosmossdk.io/math"
+	avstypes "github.com/ExocoreNetwork/exocore/x/avs/types"
+	"github.com/ExocoreNetwork/exocore/x/feedistribution/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// Based on the epoch, AllocateTokens performs reward and fee distribution to all validators.
+func (k Keeper) AllocateTokens(ctx sdk.Context, totalPreviousPower int64) error {
+	logger := k.Logger()
+	feeCollector := k.authKeeper.GetModuleAccount(ctx, k.feeCollectorName)
+	feesCollectedInt := k.bankKeeper.GetAllBalances(ctx, feeCollector.GetAddress())
+	feesCollected := sdk.NewDecCoinsFromCoins(feesCollectedInt...)
+
+	// transfer collected fees to the distribution module account
+	if err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, k.feeCollectorName, types.ModuleName, feesCollectedInt); err != nil {
+		return err
+	}
+
+	feePool := k.GetFeePool(ctx)
+	if totalPreviousPower == 0 {
+		feePool.CommunityPool = feePool.CommunityPool.Add(feesCollected...)
+		k.SetFeePool(ctx, feePool)
+		return nil
+	}
+	logger.Info("Allocate tokens to all validators", "feesCollected amount is ", feesCollected)
+	// calculate fraction allocated to exocore validators
+	remaining := feesCollected
+	communityTax, err := k.GetCommunityTax(ctx)
+	if err != nil {
+		return err
+	}
+	feeMultiplier := feesCollected.MulDecTruncate(math.LegacyOneDec().Sub(communityTax))
+
+	// allocate tokens proportionally to voting power of different validators
+	// TODO: Consider parallelizing later
+	allValidators := k.StakingKeeper.GetAllExocoreValidators(ctx) // GetAllValidators(suite.Ctx)
+	for i, val := range allValidators {
+		pk, err := val.ConsPubKey()
+		if err != nil {
+			logger.Error("Failed to deserialize public key; skipping", "error", err, "i", i)
+			continue
+		}
+		validatorDetail, found := k.StakingKeeper.ValidatorByConsAddrForChainID(
+			ctx, sdk.GetConsAddress(pk), avstypes.ChainIDWithoutRevision(ctx.ChainID()),
+		)
+		if !found {
+			logger.Error("Operator address not found; skipping", "consAddress", sdk.GetConsAddress(pk), "i", i)
+			continue
+		}
+		if totalPreviousPower == 0 {
+			return nil
+		}
```

Comment:
```suggestion
```
This can be removed since we already checked this outside the loop. Or did you mean to check `val.Power == 0` ?